cmd/update-kubernetes-release-version/main.go

```diff
@@ -0,0 +1,197 @@
+//
+// Purpose: to update the latest version of Kubernetes as it's known by this repo
+// Usage: go run cmd/update-kubernetes-release-version/main.go
+//
+// Behaviours
+// - if the latest version matches what's in the first entry of the resources/coverage/releases.yaml file, the date of it's release will be set and a new unreleased version will be pushed to the first entry
+// - if the latest version is less than the first entry of the resources/coverage/releases.yaml file, nothing will happen
+//
+// TODO quick refactor into functions
+// TODO add tests
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"text/template"
+	"time"
+
+	semver "github.com/hashicorp/go-version"
+	"sigs.k8s.io/yaml"
+)
+
+type Release struct {
+	Version     string `json:"version"`
+	ReleaseDate string `json:"release_date"`
+}
+
+type GitHubRelease struct {
+	CreatedAt string `json:"created_at"`
+}
+
+func ReadFile(path string) (string, error) {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return "", err
+	}
+	return string(content), nil
+}
+
+func WriteFile(path string, content string) error {
+	_, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(path, []byte(content), 0644)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GetHTTPFile(uri string) (content string, resp *http.Response, err error) {
+	resp, err = http.Get(uri)
+	if err != nil {
+		return "", nil, err
+	}
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", nil, err
+	}
+	return string(body), resp, nil
+}
+
+func main() {
+	log.Println("Running update-kubernetes-release-version")
+
+	// parse resources/coverage/releases.yaml
+	releasesYAML, err := ReadFile("resources/coverage/releases.yaml")
+	if err != nil {
+		log.Fatalf("unable to read resources/coverage/releases.yaml, %v", err)
+	}
+	var releases []Release
+	err = yaml.Unmarshal([]byte(releasesYAML), &releases)
+	if err != nil {
+		log.Fatalf("failed to parse releasesYAML, %v", err)
+	}
+	fmt.Println(releases)
+
+	// download kubernetes stable.txt
+	stableVersionTxt, _, err := GetHTTPFile("https://storage.googleapis.com/kubernetes-release/release/stable.txt")
+	if err != nil {
+		log.Fatalf("failed to fetch stable.txt, %v", err)
+	}
+	fmt.Println(stableVersionTxt)
+
+	// parse the version string
+	// stableVersionTxt = "v1.26.0"
+	stableVersion, err := semver.NewSemver(stableVersionTxt)
+	if err != nil {
+		log.Fatalf("failed to parse stableVersion string, %v", err)
+	}
+	fmt.Println(stableVersion.Segments())
+
+	// contruct the major release version
+	lastMajorRelease := fmt.Sprintf("%v.%v.0", stableVersion.Segments()[0], stableVersion.Segments()[1])
+	fmt.Println(lastMajorRelease)
+
+	// get the release date for the current major version
+	req, err := http.NewRequest(http.MethodGet, "https://api.github.com/repos/kubernetes/kubernetes/releases/tags/v"+lastMajorRelease, nil)
```

Comment:
move to variable at the top (https://api.github.com/repos/kubernetes/kubernetes/releases/tags)

Reply:
Resolved in https://github.com/cncf/apisnoop/commit/75b79a6d20df9fd15be03f5da9ac1ef744fb0bf5